enterprise/server/routing/routing_action_cache_client/routing_action_cache_client.go

```diff
@@ -101,21 +102,32 @@ func (r *RoutingACClient) GetActionResult(ctx context.Context, req *repb.GetActi
 }
 
 func (r *RoutingACClient) UpdateActionResult(ctx context.Context, req *repb.UpdateActionResultRequest, opts ...grpc.CallOption) (*repb.ActionResult, error) {
-	primaryClient, _, err := r.router.GetACClients(ctx)
+	primaryClient, secondaryClient, err := r.router.GetACClients(ctx)
 	if err != nil {
 		return nil, status.InternalErrorf("Failed to get primary AC client: %s", err)
 	}
-	rsp, err := primaryClient.UpdateActionResult(ctx, req, opts...)
-	if err != nil {
-		return rsp, err
-	}
 
 	c, err := r.router.GetCacheRoutingConfig(ctx)
 	if err != nil {
+		// This should never happen, but if it does, we at least know where the
+		// primary request should go, so let's send that.
 		log.CtxWarningf(ctx, "Failed to fetch routing config: %s", err)
-		return rsp, nil
+		return primaryClient.UpdateActionResult(ctx, req, opts...)
 	}
 	singleRandValue := rand.Float64()
+	dualWrite := singleRandValue < c.GetDualWriteFraction()
+	if dualWrite && secondaryClient != nil {
+		var wg sync.WaitGroup
+		wg.Go(func() { secondaryClient.UpdateActionResult(ctx, req, opts...) })
+		defer wg.Wait()
+		return primaryClient.UpdateActionResult(ctx, req, opts...)
+	}
+
+	rsp, err := primaryClient.UpdateActionResult(ctx, req, opts...)
+	if err != nil {
+		return rsp, err
```

Comment:
should we be comparing the responses here?

Reply:
ehh - we have a read validation path, so i think logging errors on the secondary write oughta be good enough?  lemme know if you feel strongly about this.